### ISSUE TITLE
🐛 Bug: Ersilia Python API Fetch refactoring

### ISSUE THREAD

**Issue by GemmaTuron** [ersilia-os/ersilia#1449]
### Describe the bug.

I am trying to use Ersilia as a Python Package. The functions work if the model is fetched but it is unable to actually fetch the model from DockerHub.
As a template I am using the functions prepared for ChemSampler: https://github.com/ersilia-os/chem-sampler/blob/main/chemsampler/core/base.py

The only message I get at fetching time is:

```
>>> from ersilia.hub.fetch.fetch import ModelFetcher
>>> mf = ModelFetcher(force_from_dockerhub=True)
13:10:57 | DEBUG    | Initialized with URL: None
13:10:57 | DEBUG    | Getting model source
13:10:57 | DEBUG    | Model getting fetched from DockerHub
>>> mf.fetch("eos4avb")
<coroutine object ModelFetcher.fetch at 0x7998692cf890>
```

### Operating environment

Ubuntu 24.04 LTS

**Comment by Abellegese** [comment]
Hey @GemmaTuron its `async` function and needed to be awaited or run using `asyncio` just put it like this

```python
asyncio.run(mf.fetch("eos4avb"))
```

**Comment by GemmaTuron** [comment]
ok, this works @Abellegese 

When was this change introduced? Previously asyncio  was not required, and why the other functions like serve and run do not need it? And important as well, where is this documented?

**Comment by Abellegese** [comment]
Okay @GemmaTuron nice question.

This change was the first introduced before even improving the run command. I started with the fetch command making them light and non-block and concurrent (look this issue https://github.com/ersilia-os/ersilia/issues/1299). This process is intensive and I used async + concurrency to make them resource efficient and non-blocking. Serve was optimized so I had not touch it.

Does it makes sense?

**Comment by GemmaTuron** [comment]
This makes sense @Abellegese thanks, but these changes are not acceptable without a user-wide warning and proper documentation.  If you change a function that is user-facing it needs to be stated everywhere. At the moment, both ZairaChem and ChemSampler will fail because of this change.


Our current documentation on the python package is here: https://ersilia.gitbook.io/ersilia-book/ersilia-model-hub/local-inference

Can you please clarify what happens if I only run, as suggested:
```
from ersilia import ErsiliaModel
mdl = ErsiliaModel("retrosynthetic-accessibility")
mdl.serve()
```
Will the serve automatically try to fetch the model, and will it do it properly using the asyncio function ?

**Comment by GemmaTuron** [comment]
In ersilia/coremodel.py 
I see the following:
```
            if do_fetch:
                fetch = importlib.import_module("ersilia.hub.fetch.fetch")
                mf = fetch.ModelFetcher(
                    config_json=self.config_json, credentials_json=self.credentials_json
                )
                mf.fetch(self.model_id)
```

**Comment by Abellegese** [comment]
I don't think you have the latest ersilia.

**Comment by GemmaTuron** [comment]
I pip installed it, which means the asyncio was already required in the latest release but was not there actually?

**Comment by Abellegese** [comment]
Yeah but just `pull` rebase from ersilia. I think this change has been made 3 weeks ago.

**Comment by GemmaTuron** [comment]
If that is the case we need to be more careful about the pip relases because we cannot release code that will not work. In
`ersilia v0.1.39` the ModelFetcher does not work, which makes me think that this asyncio bit was not there but was already required? have not looked deeply into it yet.

**Comment by Abellegese** [comment]
Honestly @GemmaTuron I don't know the release pipeline work. But this change has been made long ago. 

Okay I will make sure this does not happen. I will try my best.

**Comment by GemmaTuron** [comment]
I am reopening the issue to track the maintenance of this feature. @DhanshreeA is the ErsiliaModel module tested consistently in CI/CD workflows? It seems it might not be given that the v0.1.39 is failing. Can we include it in the testing before a new version is released?

**Comment by DhanshreeA** [comment]
However there is no test for using ErsiliaModel when a model isn't fetched (ie testing that serve automatically fetches it).

**Comment by GemmaTuron** [comment]
After giving some thought, this is how the Python API should look like:

```
from ersilia import ErsiliaModel
mdl = ErsiliaModel("retrosynthetic-accessibility")
mdl.fetch()
mdl.serve()
mdl.run(input, output)
```
This means we have to add a small refactoring on the fetch function creating a wrapper for the asyncio.run(fetch) and of course updating the CI/CD tests

**Comment by GemmaTuron** [comment]
I am reopening this issue because it is not clear to me how is the command running. Is the documentation up to date? Do we need to fetch the model or not? Currently the documentation says we do not need to. 
I need to confirm this for ZairaChem ASAP as I have the models fetched but probably the code will not work from scratch.

This is what I am currently using:


```
    def load_model(self):
        self.model = ErsiliaModel(
            model=self.model_id,
            save_to_lake=False,
            service_class="pulled_docker",
            fetch_if_not_available=True,
        )

    def run(self,input_csv, output_h5):
        self.model.serve()
        self.model.run(input=input_csv, output=output_h5)
        self.model.close()
```

**Comment by GemmaTuron** [comment]
Hi @DhanshreeA and @miquelduranfrigola 

Design question for the Python API vs the CLI:

The commands in the Ersilia CLI are:
```
ersilia fetch [MODEL_ID]
ersilia serve [MODEL_ID]
ersilia run -i
```

but we can also do simply:
```
ersilia serve [MODEL_ID]
ersilia run -i
```
And the serve command will try to fetch the model from DockerHub or S3 or GitHub (in that order if I am not confused)

The commands for the python API are:

from ersilia import ErsiliaModel
em = ErsiliaModel("[MODEL_ID]")
em.serve()
em.run()

The instantiation of the class ErsiliaModel automatically fetches the model from_Dockerhub unless specified in the `service_class` parameter. 

So, in the Python API the fetch does not automatically occur in the serve but at class instantiation, whereas in the CLI it occurs in the serve. 
It makes sense, but I just want to make sure this is what is happening to document it appropriately
